### PR TITLE
Work around Windows Activation Context problems

### LINKED ISSUE #87: pyimport results in Runtime Error
When answering [a qustion](http://stackoverflow.com/questions/25155205/ijulia-notebook-ipython-files) on stack overflow I noticed that if I import a part of IPython with `@pyimport` up pops a alert windows alerting of a runtime error. 
```
using PyCall
@pyimport IPython.nbformat.current as ipyt
```
The alert windows says 
```
Runtime Error!

Program C:\Users\admin\AppData\Local\Julia-0.3.0-rc2\bin\julia.exe

R6034
An application has made an attempt to load the C runtime library incorrectly.
Please contact the application's support team for more information.
```
Interestingly enough the import still works and I can use `ipyt` to access the notebook parsing function of IPython.

This is on Windows 8.1 Pro 64 bit.
```
versioninfo()

Julia Version 0.3.0-rc2+12
Commit 3ddbaa1 (2014-08-06 22:29 UTC)
Platform Info:
  System: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i5-2500K CPU @ 3.30GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Sandybridge)
  LAPACK: libopenblas
  LIBM: libopenlibm
  LLVM: libLLVM-3.3
```
```
Pkg.status()

20 required packages:
 - ApproxFun                     0.0.2
 - GLFW                          1.0.0-alpha.3
 - GLM                           0.4.0
 - GLPK                          0.2.11
 - GLPKMathProgInterface         0.1.5
 - Gadfly                        0.3.4+             master
 - HypothesisTests               0.2.7
 - IJulia                        0.1.12
 - Images                        0.2.47
 - Jewel                         0.6.4
 - JuMP                          0.5.5
 - MixedModels                   0.3.9
 - Optim                         0.3.1
 - PValueAdjust                  1.0.0
 - RDatasets                     0.1.1
 - Resampling                    0.0.0
 - Rmath                         0.0.0
 - SIUnits                       0.0.2
 - TypeCheck                     0.0.3
 - Winston                       0.11.2
59 additional packages:
 - ArrayViews                    0.4.6
 - BinDeps                       0.2.14
 - Cairo                         0.2.15
 - Calculus                      0.1.4
 - Cartesian                     0.2.2
 - Codecs                        0.1.2
 - Color                         0.2.11
 - Compose                       0.3.4+             master
 - Contour                       0.0.2
 - DataArrays                    0.2.0
 - DataFrames                    0.5.7
 - DataStructures                0.3.1
 - Datetime                      0.1.6
 - Distance                      0.4.0
 - Distributions                 0.5.4
 - DualNumbers                   0.1.0
 - ForwardDiff                   0.0.2
 - GLAbstraction                 0.0.0-             master (unregistered)
 - GLPlot                        0.0.0-             master (unregistered)
 - GLWindow                      0.0.0-             master (unregistered)
 - GZip                          0.2.13
 - Graphs                        0.4.3
 - HTTPClient                    0.1.4
 - Hexagons                      0.0.2
 - ImmutableArrays               0.0.6
 - IniFile                       0.2.2
 - Iterators                     0.1.6
 - JSON                          0.3.7
 - KernelDensity                 0.0.2
 - Lazy                          0.4.1
 - LibCURL                       0.1.3
 - LibExpat                      0.0.4
 - Loess                         0.0.2
 - MathProgBase                  0.2.4
 - ModernGL                      0.0.0-             master (unregistered)
 - Mustache                      0.0.3
 - NLopt                         0.1.1
 - Nettle                        0.1.4
 - NeuralNets                    0.0.0-             master (unregistered)
 - NumericExtensions             0.6.2
 - NumericFuns                   0.2.3
 - Options                       0.2.2
 - PDMats                        0.2.3
 - PyCall                        0.4.8
 - PyPlot                        1.2.10
 - Quaternions                   0.0.4
 - REPLCompletions               0.0.1
 - React                         0.1.3
 - Reexport                      0.0.1
 - ReverseDiffSparse             0.1.2
 - SortingAlgorithms             0.0.1
 - StatsBase                     0.6.3
 - TexExtensions                 0.0.1
 - Tk                            0.2.13
 - URIParser                     0.0.2
 - URLParse                      0.0.0
 - WinRPM                        0.1.2
 - ZMQ                           0.1.13
 - Zlib                          0.1.7
```

### PR DESCRIPTION
**Pull request by Keno**
For a detailed explanation of the problems, see https://github.com/zeromq/pyzmq/issues/861.
Fixes #87.

### CODE REVIEW

File: src/PyCall.jl [hunk 1]
```diff
@@ -334,14 +334,63 @@ end
...
+    # through ctypes, which does not have this functionality. Work around
+    # this by manually activating the activation context before any call to
+    # PyImport_ImportModule, since extensions are most likely to be loaded
+    # during import.
+
+    immutable ACTIVATION_CONTEXT_BASIC_INFORMATION
+        handle::Ptr{Void}
+        dwFlags::UInt32
+    end
+    const ActivationContextBasicInformation = 1
```
Comment (stevengj) [inline src/PyCall.jl]: The [QueryActCtxW docs](https://msdn.microsoft.com/en-us/library/aa375704(v=vs.85).aspx) list this constant as "not available" ... ?

Comment (Keno) [inline src/PyCall.jl]: It's lying.

**Comment by stevengj** [comment]
Thanks for doing this, @Keno.   I'm in favor of merging this.

Is `ActivatePyActCtx` expensive compared to a typical `pyimport`?  If I have to do a whole bunch of `pyimport` calls, would it be better to have a way to do this once?  I'm hoping it is cheap enough that I can not worry.

**Comment by Keno** [comment]
> Is ActivatePyActCtx expensive compared to a typical pyimport? If I have to do a whole bunch of pyimport calls, would it be better to have a way to do this once? I'm hoping it is cheap enough that I can not worry.

We could cache the activation context lookup, which is the expensive part. Activating the context itself should be very cheap. Windows basically does it on every call to a dll that has a different context from the calling one.

**Comment by stevengj** [comment]
Can you add a test?   Maybe change [this test](https://github.com/stevengj/PyCall.jl/blob/ee0897b94268d3d380bfc810b08fa666f01a8f88/test/runtests.jl#L274-L284) to use `pyzmq` instead of `conda-api`?

**Comment by stevengj** [comment]
Caching the lookup sounds like a good idea if it is expensive.

**Comment by Keno** [comment]
I'm not sure if it is expensive, but certainly more expensive than the actual activation. I'll add code to do the caching and adjust the test as you suggested.

File: src/pyinit.jl [hunk 2]
```diff
@@ -47,6 +47,9 @@ function __init__()
             end
         end
         ccall((@pysym :Py_InitializeEx), Void, (Cint,), 0)
+
+        # Will get reinitialized properly on first use
+        is_windows() && (PyActCtx[] = C_NULL)
```
Comment (stevengj) [inline src/pyinit.jl]: Shouldn't this line be after the `end`, i.e. outside the `if !already_inited` block?

File: test/runtests.jl [hunk 3]
```diff
@@ -274,11 +274,11 @@ end
 import Conda
 if PyCall.conda
     # install the tiny conda-api package as a test
```
Comment (stevengj) [inline test/runtests.jl]: comment needs to be changed: `# import pyzmq to test PR #294`.

Will importing `pyzmq` cause problems if we run `Pkg.test("PyCall")` inside IJulia or in any other context where `ZMQ.jl` has already loaded and initialized `libzmq`?

File: src/pyinit.jl (hunk 2 above)
Comment (Keno) [inline src/pyinit.jl]: Yes

**Comment by Keno** [comment]
I tested on Linux and there was no problem with both pyzmq and ZMQ being loaded. I would test on Windows, but I can't get WinRPM to work. If it becomes a problem we can disable the tests if we find ZMQ already loaded.

File: src/pyinit.jl [hunk 4]
```diff
@@ -47,6 +47,7 @@ function __init__()
             end
         end
         ccall((@pysym :Py_InitializeEx), Void, (Cint,), 0)
+
```
Comment (stevengj) [inline src/pyinit.jl]: blank line

File: src/pyinit.jl [hunk 5]
```diff
@@ -114,4 +115,7 @@ function __init__()
             end
         end
     end
+
+    # Will get reinitialized properly on first use
+    is_windows() && (PyActCtx[] = C_NULL)
```
Comment (stevengj) [inline src/pyinit.jl]: It seems like this has to go at the top of `__init__`, not at the bottom.   (Note that `__init__` calls `pyimport` several times.)

Comment (Keno) [inline src/pyinit.jl]: Fair point.

**Comment by Keno** [comment]
Good to go?

**Pull request merged**